src/java/org/apache/cassandra/net/StartupClusterConnectivityChecker.java

```diff
@@ -48,81 +51,133 @@
             logger.warn("setting the block-for-peers timeout (in seconds) to {} might be a bit excessive, but using it nonetheless", timeoutSecs);
         long timeoutNanos = TimeUnit.SECONDS.toNanos(timeoutSecs);
 
-        return new StartupClusterConnectivityChecker(targetPercent, timeoutNanos);
+        return new StartupClusterConnectivityChecker(timeoutNanos, blockForRemoteDcs);
     }
 
     @VisibleForTesting
-    StartupClusterConnectivityChecker(int targetPercent, long timeoutNanos)
+    StartupClusterConnectivityChecker(long timeoutNanos, boolean blockForRemoteDcs)
     {
-        this.targetPercent = Math.min(100, Math.max(0, targetPercent));
+        this.blockForRemoteDcs = blockForRemoteDcs;
         this.timeoutNanos = timeoutNanos;
     }
 
     /**
      * @param peers The currently known peers in the cluster; argument is not modified.
+     * @param getDatacenterSource A function for mapping peers to their datacenter.
      * @return true if the requested percentage of peers are marked ALIVE in gossip and have their connections opened;
      * else false.
      */
-    public boolean execute(Set<InetAddressAndPort> peers)
+    public boolean execute(Set<InetAddressAndPort> peers, Function<InetAddressAndPort, String> getDatacenterSource)
     {
-        if (targetPercent == 0 || peers == null)
+        if (peers == null || this.timeoutNanos < 0)
             return true;
 
         // make a copy of the set, to avoid mucking with the input (in case it's a sensitive collection)
         peers = new HashSet<>(peers);
-        peers.remove(FBUtilities.getBroadcastAddressAndPort());
+        InetAddressAndPort localAddress = FBUtilities.getBroadcastAddressAndPort();
+        String localDc = getDatacenterSource.apply(localAddress);
 
+        peers.remove(localAddress);
         if (peers.isEmpty())
             return true;
 
-        logger.info("choosing to block until {}% of the {} known peers are marked alive and connections are established; max time to wait = {} seconds",
-                    targetPercent, peers.size(), TimeUnit.NANOSECONDS.toSeconds(timeoutNanos));
+        // make a copy of the datacenter mapping (in case gossip updates happen during this method or some such)
+        Map<InetAddressAndPort, String> datacenterMap = peers.stream()
+                                                             .collect(Collectors.toMap(k -> k, getDatacenterSource));
+        Function<InetAddressAndPort, String> getDatacenter = datacenterMap::get;
 
-        long startNanos = System.nanoTime();
+        Map<String, Set<InetAddressAndPort>> peersByDc = peers.stream()
+                                                              .collect(Collectors.groupingBy(getDatacenter,
+                                                                                             Collectors.toSet()));
+
+        if (!blockForRemoteDcs)
+        {
+            peersByDc.keySet().retainAll(Collections.singleton(localDc));
```

Comment:
This important bit of behavior is tucked away here and I only know about it because I went looking. Can you add a mostly redundant comment mentioning that you remove all but the local DC so that we only wait on nodes in the local DC?

Reply:
Ack, added the comment.